File: src/forge/prompt_log.py
import json
import threading
from datetime import datetime, timezone
from pathlib import Path

_SESSIONS_DIR = Path.home() / ".forge" / "sessions"


def log_path(session_id: str) -> Path:
    return _SESSIONS_DIR / session_id / "logs" / "prompts.log"


class PromptLogger:
    def __init__(self, session_id: str) -> None:
        self._file = log_path(session_id)
        self._lock = threading.Lock()
        self._file.parent.mkdir(parents=True, exist_ok=True)

    def log(
        self,
        agent: str,
        tier: str,
        model: str,
        messages: list[dict],
        response: str,
        tokens_in: int,
        tokens_out: int,
        cost_usd: float,
        tools_called: list[str] | None = None,
    ) -> None:
        user_prompt = ""
        for m in reversed(messages):
            if m.get("role") == "user" and isinstance(m.get("content"), str):
                user_prompt = m["content"]
                break

        entry: dict = {
            "ts": datetime.now(timezone.utc).isoformat(),
            "agent": agent,
            "tier": tier,
            "model": model,
            "tokens_in": tokens_in,
            "tokens_out": tokens_out,
            "cost_usd": cost_usd,
            "user_prompt": user_prompt,
            "response": response,
        }
        if tools_called:
            entry["tools_called"] = tools_called

        with self._lock:
            with self._file.open("a") as f:
                f.write(json.dumps(entry) + "\n")

File: src-ts/promptLog.ts
import * as fs from "fs";
import * as path from "path";
import * as os from "os";
import type { CoreMessage } from "ai";

const SESSIONS_DIR = path.join(os.homedir(), ".forge", "sessions");

export function logPath(sessionId: string, sessionsDir = SESSIONS_DIR): string {
  return path.join(sessionsDir, sessionId, "logs", "prompts.log");
}

interface LogEntry {
  agent: string;
  tier: string;
  model: string;
  messages: CoreMessage[];
  response: string;
  tokensIn: number;
  tokensOut: number;
  costUsd: number;
  toolsCalled?: string[];
}

export class PromptLogger {
  private filePath: string;

  constructor(sessionId: string, sessionsDir = SESSIONS_DIR) {
    this.filePath = logPath(sessionId, sessionsDir);
    fs.mkdirSync(path.dirname(this.filePath), { recursive: true });
  }

  log(entry: LogEntry): void {
    let userPrompt = "";
    for (let i = entry.messages.length - 1; i >= 0; i--) {
      const m = entry.messages[i];
      if (m.role === "user" && typeof m.content === "string") {
        userPrompt = m.content;
        break;
      }
    }
    const record: Record<string, unknown> = {
      ts: new Date().toISOString(),
      agent: entry.agent,
      tier: entry.tier,
      model: entry.model,
      tokens_in: entry.tokensIn,
      tokens_out: entry.tokensOut,
      cost_usd: entry.costUsd,
      user_prompt: userPrompt,
      response: entry.response,
    };
    if (entry.toolsCalled?.length) record["tools_called"] = entry.toolsCalled;
    fs.appendFileSync(this.filePath, JSON.stringify(record) + "\n");
  }
}
